common/include/camera/ICameraControl.hpp

```diff
@@ -0,0 +1,312 @@
+#ifndef _NBL_I_CAMERA_CONTROLLER_HPP_
+#define _NBL_I_CAMERA_CONTROLLER_HPP_
+
+#include "IProjection.hpp"
+#include "../ICamera.hpp"
+#include "CVirtualCameraEvent.hpp"
+#include "glm/glm/ext/matrix_transform.hpp" // TODO: TEMPORARY!!! whatever used will be moved to cpp
+#include "glm/glm/gtc/quaternion.hpp"
+#include "nbl/builtin/hlsl/matrix_utils/transformation_matrix_utils.hlsl"
+
+// TODO: DIFFERENT NAMESPACE
+namespace nbl::hlsl 
+{
+
+template<ProjectionMatrix T = float64_t4x4>
+class ICameraController : virtual public core::IReferenceCounted
+{
+public:
+    using projection_t = typename IProjection<typename T>;
+
+    enum VirtualEventType : uint8_t
+    {
+        // Strafe forward
+        MoveForward = 0,
+
+        // Strafe backward
+        MoveBackward,
+
+        // Strafe left
+        MoveLeft,
+
+        // Strafe right
+        MoveRight,
+
+        // Strafe up
+        MoveUp,
+
+        // Strafe down
+        MoveDown,
+
+        // Tilt the camera upward (pitch)
+        TiltUp,
+
+        // Tilt the camera downward (pitch)
+        TiltDown,
+
+        // Rotate the camera left around the vertical axis (yaw)
+        PanLeft,
+
+        // Rotate the camera right around the vertical axis (yaw)
+        PanRight,
+
+        // Roll the camera counterclockwise around the forward axis (roll)
+        RollLeft,
+
+        // Roll the camera clockwise around the forward axis (roll)
+        RollRight,
+
+        // Reset the camera to the default state
+        Reset,
+
+        EventsCount
+    };
+
+    //! Virtual event representing a manipulation
+    struct CVirtualEvent
+    {
+        using manipulation_encode_t = float64_t;
+
+        VirtualEventType type;
+        manipulation_encode_t value;
```

Comment:
can call it `magnitude` for clarity if you want

Reply:
also wouldn't the events belong to an `IGimbal` because they're for gimbals?

Reply:
> can call it magnitude for clarity if you want

resolved

> also wouldn't the events belong to an IGimbal because they're for gimbals?

if we have generic virtual camera events like now then I would say no, also what `IGimbal` would be? Do we expect any derived classes for which it would be base? Imo any type of camera can be currently implemented with `CGimbal` instances, so for me `IGimbal` doesn't make much sense at least I don't see what we would store there. I have also moved events to new `CVirtualGimbalEvent` outside templates in recent commit

Reply:
actually let's create IGimbal which is interface for any object not just camera and add scale events